lib/authlogic/session/password.rb

```diff
@@ -154,12 +154,22 @@ def credentials
         # hash form.
         def credentials=(value)
           super
-          values = parse_param_val(value) # add strong parameters check
-
-          if values.first.is_a?(Hash)
-            values.first.with_indifferent_access.slice(login_field, password_field).each do |field, value|
-              next if value.blank?
-              send("#{field}=", value)
+          values = value.is_a?(Array) ? value.first : value
+
+          if values.respond_to? :keys
+            values = values.with_indifferent_access if values.is_a?(Hash)
+            values = values.slice(login_field, password_field)
+            missing_values = [login_field.to_s, password_field.to_s] - values.keys.map(&:to_s)
+
+            if missing_values.empty?
+              values.each do |field, value|
+                next if value.blank?
+                send("#{field}=", value)
+              end
+            else
+              raise ArgumentError.new "The following parameter(s) were missing from the supplied "\
+                                      "credentials: #{missing_values.join(', ')}. "\
+                                      "Have you supplied or permitted these parameters?"
```

Comment:
1. I like how you mention which parameters are missing. That's helpful.
1. I don't think the last sentence ("Have you supplied ...?") is necessary.
1. Please extract a constant.

```
module Password
  E_PASSWORD_PARAMS_MISSING = <<-EOS
    The following parameter(s) were missing .. %s ...
  EOS    
  # ..
  raise ArgumentError, format(E_PASSWORD_PARAMS_MISSING, missing_values.join(', '))
```

Reply:
Will do. For 2, I was trying to provide a hint relating to strong parameters but I'll pull it out as at least we still give a bit of a clue to the user.